### ISSUE TITLE
Interval analysis gives unsound saftey proof

### ISSUE THREAD

**Issue by Novak756** [esbmc/esbmc#1565]
Hi, for the following program
```c
extern void __VERIFIER_error(void);

extern long __VERIFIER_nondet_long(void);
extern _Bool __VERIFIER_nondet_bool(void);

int main(){
        _Bool b1 = __VERIFIER_nondet_bool();
        _Bool b2 = __VERIFIER_nondet_bool();
        long l1 = __VERIFIER_nondet_long();
        long l2 = __VERIFIER_nondet_long();
        if(b2){
                if(!(((b1) || ((l2) == (0))) && (!(b2)))){
                        if(!(((b2) || ((l1) == (0))) && (!(b1)))){
                        __VERIFIER_error();
                        }
                }
        }
    return 0;
}
```
running esbmc with `--interval-analysis` enabled returns `VERIFICATION_SUCCESSFUL`.
Running without flags gives the correct result (`VERIFICATION_FAILED`).

Output with `--interval-analysis --interval-analysis-dump`: [res.txt](https://github.com/esbmc/esbmc/files/13801107/res.txt)
OS: Docker Ubuntu 20.04 on WSL 2 Ubuntu
Version: Commit 9b63b634905094e17495bebf5a1fd4269a376380

**Comment by lucasccordeiro** [comment]
@rafaelsamenezes: could you please check this issue?

**Comment by rafaelsamenezes** [comment]
I'm still investigating, but some notes:

- The intervals seems correct, the only interval computed is `1 <= b2` after the first condition.
- The issue seems to be in the boolean interval propagation.

```
IF !b2 THEN GOTO 1
IF !(!(((_Bool)((signed int)b1) || l2 == 0) && !b2)) THEN GOTO 1
ASSERT ((_Bool)((signed int)b2) || l1 == 0) && !b1
```

Is being converted to 

```
IF !b2 THEN GOTO 1
IF !((signed int)b1 == 0) || l2 == 0 THEN GOTO 1
ASSERT !b1
```

In the assertion, `l1 == 0` is being wrongly removed.

**Comment by rafaelsamenezes** [comment]
Apparently the issue is related to typecasting with Integer intervals. Changing to the wrapped domains (with an extra patch) changes the code to

```
IF !b2 THEN GOTO 1
ASSERT !b1
```

The middle IF was removed.

**Comment by lucasccordeiro** [comment]
Closed via #1625.

**Comment by lucasccordeiro** [comment]
@Novak756: Thank you very much for this report. Please let us know if you find further issues.
